janusgraph-core/src/main/java/org/janusgraph/graphdb/query/graph/GraphCentricQueryBuilder.java

```diff
@@ -258,19 +263,24 @@ public GraphCentricQuery constructQueryWithoutProfile(final ElementCategory resu
         if (limit == 0) return GraphCentricQuery.emptyQuery(resultType);
 
         if (globalConstraints.isEmpty()) {
-            globalConstraints.add(constraints);
+            globalConstraints.add(Collections.singletonList(constraints));
         }
         //Prepare constraints
         final MultiCondition<JanusGraphElement> conditions;
-        if (this.globalConstraints.size() == 1) {
-            conditions = QueryUtil.constraints2QNF(tx, globalConstraints.get(0));
+        if (this.globalConstraints.size() == 1 && this.globalConstraints.get(0).size() == 1) {
+            conditions = QueryUtil.constraints2QNF(tx, globalConstraints.get(0).get(0));
             if (conditions == null) return GraphCentricQuery.emptyQuery(resultType);
         } else {
-            conditions = new Or<>();
-            for (final List<PredicateCondition<String, JanusGraphElement>> child : this.globalConstraints){
-                final And<JanusGraphElement> localconditions = QueryUtil.constraints2QNF(tx, child);
-                if (localconditions == null) return GraphCentricQuery.emptyQuery(resultType);
-                conditions.add(localconditions);
+            if (this.globalConstraints.size() == 1) {
+                conditions = constructOrCondition(this.globalConstraints.get(0));
+                if (conditions == null) return GraphCentricQuery.emptyQuery(resultType);
+            } else {
+                conditions = new And<>();
```

Comment:
just trying to understand: seems this handles the case `and(or(and(a, b), c), d)`. What about `or(and(or(and(a, b), c), d), e)`? Does handle this case as well and if so, how does it handle it?

Reply:
I believe a query like that doesn't get handled here because the `HasStepFolder.foldInHasContainer()` call in the `JanusGraphStepStrategy` doesn't do anything because it's not considered valid. The code below short circuits in `HasStepFolder.foldInHasContainer()`

```
if (!validFoldInHasContainer(child.getStartStep(), false)){
    return;
}
```

Reply:
Just verified that you are correct.